### ISSUE TITLE
Uncomment root-level exports in shared/__init__.mojo

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#3759]
All the `from .core.layers import Conv2dLayer, ReLULayer, ...` style imports in `shared/__init__.mojo` are commented out with the note 'pending full layer implementation'. Now that the API names are correct and many layers are implemented, these commented-out imports should be evaluated one by one and activated where the underlying modules are ready. This would enable the convenient `from shared import Conv2dLayer` usage pattern documented in the module docstring.

_Follow-up from #3222_

**Comment by mvillmow** [comment]
# Implementation Plan

Now I have everything I need. Here's the implementation plan:

---

# Implementation Plan: Issue #3759 — Uncomment Root-Level Exports in `shared/__init__.mojo`

## 1. Objective

Evaluate each commented-out import in `shared/__init__.mojo` (lines 54–95) and activate those whose underlying implementations are confirmed ready. This enables the `from shared import <Symbol>` convenience pattern documented in the module docstring.

---

## 2. Approach

The key constraint from prior learnings and the file itself (line 141–148) is the **Mojo v0.26.1 re-export chain limitation**: `from shared import Linear` fails even when `shared/core/__init__.mojo` re-exports `Linear`. This means we must use **absolute leaf-module paths** (e.g., `from shared.core.layers.linear import Linear`) rather than relative or chained paths.

**Strategy:**

1. For each commented-out import group, grep the codebase to verify the symbol exists at the leaf module level
2. Categorize: Ready (uncomment with absolute path), Name mismatch (annotate), Not implemented (keep commented with note)
3. Activate only confirmed-ready symbols
4. Update the Public API Table in the docstring
5. Add integration tests to `tests/shared/integration/test_packaging.mojo`

**Pre-analysis of commented groups** (lines 54–95):

| Group | Symbols | Status |
|-------|---------|--------|
| Core layers | `Linear, Conv2D, ReLU, MaxPool2D, Dropout, Flatten` | Partially ready — `Linear`, `Conv2dLayer`, `ReLULayer`, `DropoutLayer` exist; `Conv2D`/`ReLU`/`Dropout` are aliases needed; `MaxPool2D`, `Flatten` not in layers/ |
| Core activations | `relu, sigmoid, tanh, softmax` | Need to verify in `shared/core/activation.mojo` |
| Core module system | `Module, Sequential` | Need to verify `shared/core/module.mojo` and `shared/core/sequential.mojo` |
| Core tensors | `Tensor, zeros, ones, randn` | Need to verify in `shared/core/extensor.mojo` |
| Training optimizers | `SGD, Adam, AdamW` | Need to verify in `shared/training/optimizers/` |
| Training schedulers | `StepLR, CosineAnnealingLR` | Need to verify in `shared/training/schedulers/` |
| Training callbacks | `EarlyStopping, ModelCheckpoint` | Need to verify |
| Training loops | `train_epoch, validate_epoch` | Need to verify |
| Data components | `TensorDataset, ImageDataset, DataLoader` | Need to verify |
| Utils | `Logger, plot_training_curves` | Need to verify |

---

## 3. Files to Create

None — all work is modifications to existing files.

---

## 4. Files to Modify

### `shared/__init__.mojo`

**Changes:**
- Replace each commented import block with either:
  - An active `from shared.<leaf.module> import <Symbol>` line (if ready)
  - A comment with `# NOT YET IMPLEMENTED — see #<issue>` (if not ready)
  - A `comptime <Alias> = <StructName>` alias where the API name differs from the struct name (e.g., `comptime ReLU = ReLULayer`)
- Update the Public API Table (lines 152–165) to reflect activated symbols
- Update the note on line 167 from "pending full layer implementation" to reflect current state

**Example activations (based on confirmed structs):**

```mojo
# Core layers (confirmed in shared/core/layers/)
from shared.core.layers.linear import Linear
from shared.core.layers.conv2d import Conv2dLayer
comptime Conv2D = Conv2dLayer  # alias to match documented API name
from shared.core.layers.relu import ReLULayer
comptime ReLU = ReLULayer
from shared.core.layers.dropout import DropoutLayer
comptime Dropout = DropoutLayer
from shared.core.layers.batchnorm import BatchNorm2dLayer
comptime BatchNorm2d = BatchNorm2dLayer
# MaxPool2D, Flatten — NOT YET IMPLEMENTED in shared/core/layers/
```

### `tests/shared/integration/test_packaging.mojo`

**Changes:**
- Add a new test function `test_layer_root_level_imports()` verifying that `Linear`, `Conv2D`/`Conv2dLayer`, `ReLU`/`ReLULayer`, `Dropout`/`DropoutLayer`, `BatchNorm2d`/`BatchNorm2dLayer` are importable from `shared`
- Add to the `main()` function to run the new test
- Follow existing test patterns: import inside test function, use `assert_true`/`assert_equal`, print `✓ <test name> passed`

---

## 5. Implementation Order

1. **Grep each candidate symbol** to confirm existence at the leaf module level:
   - `grep -rn "^struct Linear\b" shared/core/`
   - `grep -rn "^fn relu\b\|^fn sigmoid\b\|^fn tanh\b\|^fn softmax\b" shared/core/activation.mojo`
   - `grep -rn "^struct Sequential\b\|^struct Module\b" shared/core/`
   - `grep -rn "^struct SGD\b\|^struct Adam\b\|^struct AdamW\b" shared/training/`
   - Similar for schedulers, callbacks, loops, data, utils

2. **Categorize each symbol** into: Activate / Alias-needed / Not-ready

3. **Edit `shared/__init__.mojo`**:
   - Uncomment confirmed-ready imports, rewriting as absolute-path imports
   - Add `comptime` aliases for name mismatches (e.g., `ReLU = ReLULayer`)
   - Annotate remaining commented lines with `# NOT YET IMPLEMENTED` or `# NAME MISMATCH: actual name is X`
   - Update the Public API Table

4. **Edit `tests/shared/integration/test_packaging.mojo`**:
   - Add `test_layer_root_level_imports()` for newly activated symbols
   - Register in `main()`

5. **Commit** with `SKIP=mojo-format` due to GLIBC mismatch:
   ```bash
   SKIP=mojo-format git commit -m "fix(shared): uncomment ready exports in __init__.mojo

   Closes #3759"
   ```

6. **Push and create PR** linked to issue #3759

---

## 6. Verification

**Local grep checks** (before editing):
```bash
grep -rn "^struct\|^fn " shared/core/layers/
grep -rn "^fn relu\|^fn sigmoid\|^fn tanh\|^fn softmax" shared/core/activation.mojo
grep -rn "^struct SGD\|^struct Adam" shared/training/optimizers/
```

**CI validation:**
- Pre-commit hooks pass (markdownlint, trailing-whitespace, YAML, etc.)
- `just test-mojo` passes in CI Docker (GLIBC 2.35)
- `just test-group tests/shared/integration "test_packaging*.mojo"` passes
- Packaging integration tests (all 12 existing + new layer test) pass

**Manual smoke test** (inside Docker):
```bash
docker run --rm ghcr.io/homericintelligence/projectodyssey:main \
  mojo -c "from shared import Linear; print('Linear import OK')"
```

---

## 7. Skills Used

**Invoked during planning:**
- `Explore` subagent — explored `shared/__init__.mojo`, `shared/core/layers/`, and `tests/shared/integration/` to map existing struct names, commented imports, and test patterns

**Team knowledge base skills referenced (from Prior Learnings):**
- `review-commented-imports` — direct match; audit pattern for commented-out imports, determine ready vs pending, uncomment ready ones
- `mojo-package-reexport` — re-exporting symbols at higher-level import paths; informed the absolute-path import strategy and `comptime` alias pattern
- `enable-todo-tests` — enabling commented-out exports when implementation exists; informed the test addition step
- `mojo-subpackage-import-tests` — adding tests to verify both parent-package and direct sub-package import paths; informed the new `test_layer_root_level_imports()` test design
- `mojo-module-docstring-limitation` — documenting symbols that can't be re-exported due to Mojo limitations; informed how to annotate the remaining commented lines

---
*Generated by Claude Code Planner*
